### PR TITLE
Hue fade in

### PR DESCRIPTION
**Pull request by cbellier**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [ ] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Implements the possibility to create a scene with an action which turns on a Hue light progressively.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,4 +16,7 @@ coverage
 # VSCode
 .vscode
 .tmp
-.DS_Store
\ No newline at end of file
+.DS_Store
+
+# Idea
+**/.idea/*
```
Comment (atrovato) [inline .gitignore:22]: What is easy to configure it on idea IDE?
You may prepare a developer installation guide, if your memory is fresh enough.

File: front/src/components/boxs/device-in-room/DeviceRow.jsx [hunk 2]
```diff
@@ -24,7 +24,7 @@ const DeviceRow = ({ children, ...props }) => {
       />
     );
   }
-  if (props.deviceFeature.type === 'dimmer') {
+  if (['dimmer', 'brightness'].includes(props.deviceFeature.type)) {
```
Comment (atrovato) [inline front/src/components/boxs/device-in-room/DeviceRow.jsx]: Can we extract this to a file const at top level?

Comment (atrovato) [reply front/src/components/boxs/device-in-room/DeviceRow.jsx]: And use DEVICE_FEATURE_TYPES const instead of hard coded strings?

File: front/src/components/boxs/device-in-room/EditDeviceInRoom.jsx [hunk 3]
```diff
@@ -39,8 +39,8 @@ class EditDeviceInRoom extends Component {
             value: feature.selector,
             label: getDeviceFeatureName(this.context.intl.dictionary, device, feature)
           };
-          // for now, we only supports binary on/off and sensors
-          if (feature.read_only || feature.type === DEVICE_FEATURE_TYPES.LIGHT.BINARY) {
+          // for now, we only supports binary on/off, brightness and sensors
+          if (feature.read_only || [DEVICE_FEATURE_TYPES.LIGHT.BINARY, DEVICE_FEATURE_TYPES.LIGHT.BRIGHTNESS].includes(feature.type)) {
```
Comment (atrovato) [inline front/src/components/boxs/device-in-room/EditDeviceInRoom.jsx]: Can we extract this in a const a file top level? For more readability and reusability.

File: front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      });
+      const deviceOptions = devices.map(device => ({
+        value: device.selector,
+        label: device.name
+      }));
+      await this.setState({ deviceOptions });
+      this.refreshSelectedOptions(this.props);
+      return deviceOptions;
+    } catch (e) {
+      console.log(e);
```
Comment (atrovato) [inline front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx]: Accoring to eslint warn, can you `console.error` instead?

File: front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+          min="0"
+          max="100"
+        />
+        <div className="row">
+          <div className="col-md-6">
+            <Localizer>
+              <input
+                type="text"
+                className="form-control"
+                value={parameters ? parameters.durationValue : undefined}
```
Comment (atrovato) [inline front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx:121]: To let you know, you case use `get-value` (by imorting it `import 'get' from 'get-value'`) and do
`get(parameters, 'durationValue')`

or if you need default value `get(parameters, 'durationUnit', { default: 'sesonds'})`

https://www.npmjs.com/package/get-value

But works well as written now.

File: server/package.json [hunk 6]
```diff
@@ -91,7 +91,7 @@
     "semver": "^6.1.1",
     "sequelize": "^4.42.0",
     "sequelize-cli": "^5.5.1",
-    "sqlite3": "^4.0.6",
+    "sqlite3": "^4.2.0",
```
Comment (atrovato) [inline server/package.json]: Sure?

File: server/utils/constants.js [hunk 7]
```diff
@@ -260,6 +262,7 @@ const DEVICE_FEATURE_TYPES = {
   LIGHT: {
     BINARY: 'binary',
     BRIGHTNESS: 'brightness',
+    FADE_IN: 'fade-in',
```
Comment (atrovato) [inline server/utils/constants.js:304]: What if fade out? Should be propose a 'fade' with in / out options?

File: server/package.json (hunk 6 above)
Comment (cbellier) [reply server/package.json]: rolled back

File: front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx (hunk 5 above)
Comment (cbellier) [reply front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx:121]: I always prefer to use Vanilla JS instead of some module if the cost of it is acceptable. Would love to have elvis operator though.

File: front/src/components/boxs/device-in-room/EditDeviceInRoom.jsx (hunk 3 above)
Comment (cbellier) [reply front/src/components/boxs/device-in-room/EditDeviceInRoom.jsx]: Done with rebase

File: front/src/components/boxs/device-in-room/DeviceRow.jsx (hunk 2 above)
Comment (cbellier) [reply front/src/components/boxs/device-in-room/DeviceRow.jsx]: Done with rebase

**Comment by cbellier** [comment]
[PR](https://github.com/GladysAssistant/v4-website/pull/78/files) for the documentation

File: front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx (hunk 4 above)
Comment (cbellier) [reply front/src/routes/scene/edit-scene/actions/LightFadeInParams.jsx]: Done

File: server/utils/constants.js (hunk 7 above)
Comment (cbellier) [reply server/utils/constants.js:304]: Maybe fade out can be developped as part of another feature

File: server/lib/scene/scene.actions.js [hunk 8]
```diff
@@ -91,27 +135,27 @@ const actionsFunc = {
...
         default:
-          throw new Error(`Unit ${action.unit} not recognized`);
+          throw new Error(`Unit ${action.parameters.unit} not recognized`);
       }
       setTimeout(resolve, timeToWaitMilliseconds);
     }),
   [ACTIONS.SCENE.START]: async (self, action, scope) => self.execute(action.scene, scope),
   [ACTIONS.MESSAGE.SEND]: async (self, action, scope) => {
-    const textWithVariables = Handlebars.compile(action.text)(scope);
+    const textWithVariables = Handlebars.compile(action.parameters.text)(scope);
```
Comment (Pierre-Gilles) [inline server/lib/scene/scene.actions.js:158]: I'm no sure we want to change this. Could we discuss this before ?

File: server/lib/scene/scene.actions.js [hunk 9]
```diff
@@ -91,27 +135,27 @@ const actionsFunc = {
   [ACTIONS.TIME.DELAY]: async (self, action, scope) =>
     new Promise((resolve) => {
       let timeToWaitMilliseconds;
-      switch (action.unit) {
+      switch (action.parameters.unit) {
```
Comment (Pierre-Gilles) [inline server/lib/scene/scene.actions.js:138]: I'm no sure we want to change this. Could we discuss this before ?

File: server/lib/scene/scene.actions.js [hunk 10]
```diff
@@ -58,6 +58,50 @@ const actionsFunc = {
       }
     });
   },
+  [ACTIONS.LIGHT.SET_BRIGHTNESS]: async (self, action, scope) => {
```
Comment (Pierre-Gilles) [inline server/lib/scene/scene.actions.js:61]: Is it used at the end? Could it be in a separate PR if not?

File: server/lib/scene/scene.actions.js [hunk 11]
```diff
@@ -26,7 +26,7 @@ const actionsFunc = {
       device = self.stateManager.get('device', action.device);
       deviceFeature = getDeviceFeature(device, action.feature_category, action.feature_type);
     }
-    return self.device.setValue(device, deviceFeature, action.value);
+    return self.device.setValue(device, deviceFeature, action.parameters.value);
```
Comment (Pierre-Gilles) [inline server/lib/scene/scene.actions.js:29]: I don't think we want to change this, it's not related to this PR.

File: server/lib/device/light/light.setBrightness.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * @description TurnOn a given deviceFeature.
+ * @param {Object} device - The device to change brightness of.
+ * @param {Object} deviceFeature - The deviceFeature to change brightness of.
+ * @param {number} brightness - The brightness value.
+ * @example
+ * light.setBrightness(device, 5);
+ */
+async function setBrightness(device, deviceFeature, brightness) {
```
Comment (Pierre-Gilles) [inline server/lib/device/light/light.setBrightness.js:12]: Is it used in this PR ?

File: server/lib/device/device.scenario.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * @description Control a specific device.
+ * @param {Object} device - The device to control.
+ * @param {Object} deviceFeature - The deviceFeature to control.
+ * @param {string|number} value - The new state to set.
+ * @example
+ * device.scenario(device, deviceFeature);
+ */
+async function scenario(device, deviceFeature, value) {
```
Comment (Pierre-Gilles) [inline server/lib/device/device.scenario.js:13]: I think we can find a better name for this API, why didn't you implement a fadeIn() function?

File: server/models/scene.js [hunk 14]
```diff
@@ -34,6 +50,7 @@ const actionSchema = Joi.array().items(
           value: Joi.string(),
         }),
       ),
+      parameters: parametersSchema,
```
Comment (Pierre-Gilles) [inline server/models/scene.js:53]: I'm not sure we want to wrap everything in a parameters object. Can we discuss it before?

File: server/migrations/20210124180223-scene-actions-parameters.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
+const Promise = require('bluebird');
+const db = require('../models');
+
+module.exports = {
+  up: async () => {
+    const scenes = await db.Scene.findAll({ attributes: ['id', 'actions'] });
```
Comment (Pierre-Gilles) [inline server/migrations/20210124180223-scene-actions-parameters.js:6]: I'm no sure we want to change this. Could we discuss this before ?

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
Hi!

First thanks for your PR, it's some great work and I see you spent some time to make this right 🙏 Good job!

Before looking into the code/testing this, I noticed you are making some huge API changes in Gladys. 

I'm a little bit concerned about those changed, and I think this is something that need to be discussed as it has a lot of impact (it's breaking changes, as the API changes. Not sure we want this)

I'm putting a message in the forum, we can talk about it there if you want :) 

Don't take it personally of course, this is some great work and hue fade in is definitely something we want in Gladys 😄
